_source/_posts/2021-09-20-cli-ssh-oauth-device-grant.md

```diff
@@ -0,0 +1,273 @@
+
+```
+UsePAM yes
+```
+
+Next we'll develop the authentication module. To make sure the necessary libraries are installed, we run the following commands:
+
+```sh
+apt-get update
+apt-get install build-essential libpam0g-dev libcurl4-openssl-dev libqrencode-dev libssl-dev
+```
+
+The full source code for the PAM module is available on [GitHub](https://github.com/oktadev/okta-ssh-oauth-example). For a PAM module, we need to define two interfaces:
+
+```c
+PAM_EXTERN int pam_sm_setcred( pam_handle_t *pamh, int flags, int argc, const char **argv)
+PAM_EXTERN int pam_sm_authenticate( pam_handle_t *pamh, int flags,int argc, const char **argv )
+```
+
+The second interface handles the authentication flow, which holds all our authentication logic. The code is self explanatory, so we'll only cover the high-level steps:
+
+1. First, plug in your org's urls in `deviceflow.c`. Change the url based on your org name and whether you are using the org Authentication Server or a custom Authentication Server. Also, define the client ID for the Native App where you've enabled the Device Authorization grant type.
+
+   ```c
+   #define DEVICE_AUTHORIZE_URL  "https://dev-133337.okta.com/oauth2/v1/device/authorize"
+   #define TOKEN_URL "https://dev-133337.okta.com/oauth2/v1/token"
+   #define CLIENT_ID "0oa15wulqt5yqD9FP5d7"
+   ```
+
+2. We first issue a POST call to our `authorize` endpoint on the Authorization Server. From the POST result, we can extract the `device_code` and `user_code`. The POST call is abstracted away as a call to a helper function `issuePost` which takes the URL as the first parameter and the POST data as the second parameter:
+
+   ```c
+   sprintf(postData, "client_id=%s&scope=openid profile offline_access", CLIENT_ID);
+   issuePost(DEVICE_AUTHORIZE_URL, postData);
+   ```
+
+    The helper function uses libcurl to send the POST request:
+
+   ```c
+   void issuePost(char * url, char * data) {
+      /* send all data to this callback function  */
+      curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
+      /* we pass our 'chunk' struct to the callback function */
+      curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)&chunk);
+      
+      curl_easy_setopt(curl, CURLOPT_URL, url ) ;
+      curl_easy_setopt(curl, CURLOPT_POST, 1);  /* this is a POST */
+      curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data);
+      int res = curl_easy_perform( curl ) ;
+   }
+   ```
+
+    The `WriteMemoryCallback` function saves the callback results in a `chunk` buffer. Then, we use a helper function `getValueForKey` to extract the value fields for keys `user_code`, `device_code`, and `verification_uri`. 
+
+3. We then prompt the user to login to the activate endpoint. For a more pleasant user experience, we incorporate the URL inside a QRCode. We leveraged the code from Y2Z/qr:  Text-based QR Code® generator to generate ASCII output so that it will output properly in the command shell. 
```

Comment:
```suggestion
3. We then prompt the user to login to the activate endpoint. For a more pleasant user experience, we incorporate the URL inside a QRCode. We leveraged the code from [Y2Z/qr](https://github.com/Y2Z/qr) to generate ASCII output so that it will output properly in the command shell. 
```